### PR TITLE
Make it possible to check IPython for plagiarism

### PR DESCRIPTION
**Pull request by libre-man**

### CODE REVIEW

File: jplag.frontend.cpp/src/main/java/jplag/cpp/Language.java [hunk 1]
```diff
@@ -52,15 +52,15 @@ public boolean usesIndex() {
...
+		return this.scanner.parse(dir, files);
 	}
 
 	public boolean errors() {
 		return this.scanner.getErrors();
 	}
 
 	public int noOfTokens() {
-		return jplag.cpp.CPPToken.numberOfTokens();
+		return scanner.getEndOfFileToken("").numberOfTokens();
```
Comment (olmokramer) [inline jplag.frontend.cpp/src/main/java/jplag/cpp/Language.java:63]: Should we not add a `staticNumberOfTokens` method to the CPPToken class, for symmetry with other languages?

File: jplag.frontend.csharp-1.2/src/main/java/jplag/csharp/Language.java [hunk 2]
```diff
@@ -56,7 +56,7 @@ public boolean usesIndex() {
 	}
 
 	public int noOfTokens() {
-		return jplag.csharp.CSharpToken.numberOfTokens();
+		return parser.getEndOfFileToken("").numberOfTokens();
```
Comment (olmokramer) [inline jplag.frontend.csharp-1.2/src/main/java/jplag/csharp/Language.java:59]: Same as for the CPPTokens.

File: jplag.frontend.generic/src/main/java/jplag/generic/JavascriptLanguage.java [hunk 3]
```diff
@@ -49,4 +42,9 @@ public String getShortName() {
     public int min_token_match() {
         return 8;
     }
+
+
+    public static void main(String[] args) {
```
Comment (olmokramer) [inline jplag.frontend.generic/src/main/java/jplag/generic/JavascriptLanguage.java]: This looks like it should be removed

File: jplag.frontend.ipython/src/main/java/jplag/ipython/IPythonLanguage.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+					lineNr++;
+				}
+				reader.close();
+			} catch (IOException e) {
+				System.out.println(e);
+			}
+		}
+	}
+
+	public static void main(String[] args) {
```
Comment (olmokramer) [inline jplag.frontend.ipython/src/main/java/jplag/ipython/IPythonLanguage.java]: This looks like it should be removed

File: jplag.frontend.ipython/src/main/java/jplag/ipython/IPythonParser.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        while (cells.hasNext()) {
+            int amountLines = 1;
+            int extraLines = 0;
+
+            Object item = cells.next();
+            try {
+                JSONObject cell = (JSONObject) item;
+                String cellType = cell.getString("cell_type");
+                if (cellType.equals("raw")) {
+                    amountLines = 0;
```
Comment (olmokramer) [inline jplag.frontend.ipython/src/main/java/jplag/ipython/IPythonParser.java:120]: Why reset `amountLines` here?

Comment (olmokramer) [reply jplag.frontend.ipython/src/main/java/jplag/ipython/IPythonParser.java:120]: Never mind, I thought `amountLines` was initialized outside of the while loop.

File: jplag.frontend.json/src/main/java/jplag/json/Parser.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        if (runOut)
+            return;
+        runOut = true;
+        errors++;
+        program.print("ERROR: Out of serials!", null);
+        System.out.println("jplag.json.Parser: ERROR: Out of serials!");
+    }
+
+    public static void main(String args[]) {
+        String[] ffiles = {"/home/thomas/downloads/subs/top/Exercise1.json"};
```
Comment (olmokramer) [inline jplag.frontend.json/src/main/java/jplag/json/Parser.java]: This looks like it should be removed.

File: jplag.frontend.python-3/src/main/java/jplag/python3/Parser.java [hunk 7]
```diff
@@ -16,22 +14,44 @@
...
+        ParseTreeWalker ptw = new ParseTreeWalker();
+        for (int i = 0; i < in.getChildCount(); i++) {
+            ParseTree pt = in.getChild(i);
+            ptw.walk(new JplagPython3Listener(new PythonTokenCreator(adder)), pt);
         }
+        return true;
+    }
+
+    public static void main(String args[]) {
+        String[] ffiles = {"/home/thomas/software/codegra.de/manage.py"};
```
Comment (olmokramer) [inline jplag.frontend.python-3/src/main/java/jplag/python3/Parser.java]: Looks like it should be removed.

File: jplag.frontend.python-3/src/main/java/jplag/python3/Python3TokenConstants.java [hunk 8]
```diff
@@ -35,6 +35,7 @@
     final static int DEL = 30;
     final static int WITH_BEGIN = 31;
     final static int WITH_END = 32;
+    int MAGIC_STMT = 33;
```
Comment (olmokramer) [inline jplag.frontend.python-3/src/main/java/jplag/python3/Python3TokenConstants.java:38]: Should this not be `final static` like all others?

Comment (libre-man) [reply jplag.frontend.python-3/src/main/java/jplag/python3/Python3TokenConstants.java:38]: No, because that is completely useless.

**Pull request merged**